File: cpp/questions/maximizeSquareArea.cpp
class Solution {
private:
    static constexpr int MOD = 1e9 + 7;
public:
    int maximizeSquareArea(int m, int n, vector<int>& hFences, vector<int>& vFences) {
        auto addBorderSort = [](std::vector<int>& fences, int b) -> void {
            fences.push_back(1);
            fences.push_back(b);
            ranges::sort(fences);
        };
        addBorderSort(hFences, m);
        addBorderSort(vFences, n);
        std::unordered_map<int, int> hvm;
        int maxb = 0;
        for (int& h: hFences) {
            for (int& v: vFences) {
                int curhv = h - v;
                if (hvm.count(curhv)) {
                    maxb = std::max(maxb, h-hvm[curhv]);
                } else {
                    hvm[curhv] = h;
                }
            }
        }
        return maxb==0 ? -1 : static_cast<long long>(maxb) * maxb % MOD;
    }
};

File: python/questions/maximizeSquareArea.py
class Solution:
    def maximizeSquareArea(self, m: int, n: int, hFences: List[int], vFences: List[int]) -> int:
        MOD = 10**9 + 7
        def addBorderSort(fences: List[int], border: int) -> void:
            fences.append(1)
            fences.append(border)
            fences.sort()
        addBorderSort(hFences, m)
        addBorderSort(vFences, n)
        hvd = {}
        maxb = 0;
        for h in hFences:
            for v in vFences:
                curhv = h - v
                if curhv in hvd:
                    maxb = max(maxb, h-hvd[curhv])
                else:
                    hvd[curhv] = h
        return (maxb**2)%MOD if maxb!=0 else -1 
